From 901b6bcae0e7c52d4549989c9509da7a81b9ea40 Mon Sep 17 00:00:00 2001
From: vladimir.a.ivanov <vladimir.a.ivanov@intel.com>
Date: Wed, 6 Mar 2013 11:01:24 +0700
Subject: Dalvik: missed null checks in compiled code

BZ: 90661

Null and Bound Checks optimization assume that check-cast
and instanceof bytecodes do null check. It is incorrect and
leads to missed null check in following bytecodes.
The patch removes null check flag marking these bytecodes.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-ME
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: Ic4944581e2788d83e848005f7f5055af432718cd
Orig-MCG-Change-Id: Ica4ca5cd7e3c0238d12602fa710b62b3cf292ed4
Signed-off-by: vladimir.a.ivanov <vladimir.a.ivanov@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/Dataflow.cpp |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/vm/compiler/Dataflow.cpp b/vm/compiler/Dataflow.cpp
index b197de1..da23fb6 100644
--- a/vm/compiler/Dataflow.cpp
+++ b/vm/compiler/Dataflow.cpp
@@ -125,10 +125,10 @@ int dvmCompilerDataFlowAttributes[kMirOpLast] = {
     DF_UA | DF_NULL_OBJECT_CHECK_0 | DF_CLOBBERS_MEMORY,
 
     // 1F OP_CHECK_CAST vAA, type@BBBB
-    DF_UA | DF_NULL_OBJECT_CHECK_0,
+    DF_UA,
 
     // 20 OP_INSTANCE_OF vA, vB, type@CCCC
-    DF_DA | DF_UB | DF_NULL_OBJECT_CHECK_0,
+    DF_DA | DF_UB,
 
     // 21 OP_ARRAY_LENGTH vA, vB
     DF_DA | DF_UB | DF_NULL_OBJECT_CHECK_0,
-- 
1.7.4.1

